include/fmt/format.h

```diff
@@ -278,9 +278,41 @@ FMT_END_NAMESPACE
 #  define FMT_DEPRECATED_NUMERIC_ALIGN 0
 #endif
 
+#define FMT_DATA_DIGITS                                                     \
+  {                                                                         \
+    {'0', '0'}, {'0', '1'}, {'0', '2'}, {'0', '3'}, {'0', '4'}, {'0', '5'}, \
+        {'0', '6'}, {'0', '7'}, {'0', '8'}, {'0', '9'}, {'1', '0'},         \
+        {'1', '1'}, {'1', '2'}, {'1', '3'}, {'1', '4'}, {'1', '5'},         \
+        {'1', '6'}, {'1', '7'}, {'1', '8'}, {'1', '9'}, {'2', '0'},         \
+        {'2', '1'}, {'2', '2'}, {'2', '3'}, {'2', '4'}, {'2', '5'},         \
+        {'2', '6'}, {'2', '7'}, {'2', '8'}, {'2', '9'}, {'3', '0'},         \
+        {'3', '1'}, {'3', '2'}, {'3', '3'}, {'3', '4'}, {'3', '5'},         \
+        {'3', '6'}, {'3', '7'}, {'3', '8'}, {'3', '9'}, {'4', '0'},         \
+        {'4', '1'}, {'4', '2'}, {'4', '3'}, {'4', '4'}, {'4', '5'},         \
+        {'4', '6'}, {'4', '7'}, {'4', '8'}, {'4', '9'}, {'5', '0'},         \
+        {'5', '1'}, {'5', '2'}, {'5', '3'}, {'5', '4'}, {'5', '5'},         \
+        {'5', '6'}, {'5', '7'}, {'5', '8'}, {'5', '9'}, {'6', '0'},         \
+        {'6', '1'}, {'6', '2'}, {'6', '3'}, {'6', '4'}, {'6', '5'},         \
+        {'6', '6'}, {'6', '7'}, {'6', '8'}, {'6', '9'}, {'7', '0'},         \
+        {'7', '1'}, {'7', '2'}, {'7', '3'}, {'7', '4'}, {'7', '5'},         \
+        {'7', '6'}, {'7', '7'}, {'7', '8'}, {'7', '9'}, {'8', '0'},         \
+        {'8', '1'}, {'8', '2'}, {'8', '3'}, {'8', '4'}, {'8', '5'},         \
+        {'8', '6'}, {'8', '7'}, {'8', '8'}, {'8', '9'}, {'9', '0'},         \
+        {'9', '1'}, {'9', '2'}, {'9', '3'}, {'9', '4'}, {'9', '5'},         \
+        {'9', '6'}, {'9', '7'}, {'9', '8'}, {                               \
+      '9', '9'                                                              \
+    }                                                                       \
+  }
+
 FMT_BEGIN_NAMESPACE
 namespace detail {
 
+using data_digit_pair = char[2];
+
+namespace compile_time_formatting_data {
+constexpr data_digit_pair digits[] = FMT_DATA_DIGITS;
+}
```

Comment:
Why move this out of `basic_data`?

Reply:
> Why move this out of basic_data?
> Can we avoid a macro?

I need this value at compile time as well as the entire library needs it at runtime, so to avoid having two separate definitions I added a macro definition for that. This allows me to use exactly the same value in `basic_data` and in-place. At least that was the idea, but I just don't know what is better - to create a global variable in some namespace (as right now) or to create a variable in every function that needed it.
But here is a direct answer to your question: basic data is not available at compile-time, because it is declared but not defined, only `format-inl.h` sets values in it, wherever these values defined in the static library or they defined later if {fmt} is header-only.

Reply:
I think it should be enough to make `digits` `constexpr` and move the definition to `format.h`.

Reply:
This is surely out of the scope of this PR, but moving `digits` to `format.h` is only one small piece. When someone decides to format `int` as hex then looks like `hex_digits` is used (conditionally 😒). With output coloring - `foreground_color`, `background_color`, etc. With padding - `left_padding_shifts` and `right_padding_shifts`. Needless to say, the floating-point formatting uses this data.

Anyway, I will move `digits` to `format.h` and make it `constexpr`.

Reply:
Also, this change will probably break ABI (I remember some issue about that recently), but it can be bypassed by leaving `digits` in data inside `format-inl.h`, but initializing it by `digits` inside `format.h`.

Reply:
> Also, this change will probably break ABI (I remember some issue about that recently), but it can be bypassed by leaving digits in data inside format-inl.h, but initializing it by newly created digits inside format.h.

Good point. We should definitely keep `basic_data::digits` as is.

Thinking more of it, I don't think we need `constexpr` version of `digits` at all because it's an optimization that doesn't make sense at compile time. Instead I suggest changing the implementation of `format_decimal` with a separate path guarded by `is_constant_evaluated` that does simple decimal to string conversion that doesn't require any data tables.